### PR TITLE
add self report charts

### PR DESCRIPTION
**Pull request by mikehelmick**
Towards #1928 

## Proposed Changes

* Add 2 new charts for user-reported codes
* Improve seed script so make the numbers normalized across different types + add user-report generation

image

**Release Note**

```release-note
Add charts for user-reported codes and claim stats.
```

### CODE REVIEW

File: assets/server/realmadmin/_stats_codes.html [hunk 1]
```diff
@@ -69,6 +113,65 @@
   </small>
 </div>
 
+<div class="modal fade" id="realm-user-codes-modal" data-backdrop="static" tabindex="-1" aria-hidden="true">
+  <div class="modal-dialog modal-dialog-centered">
+    <div class="modal-content">
+      <div class="modal-header">
+        <h5 class="modal-title">User Report Codes &amp; Usage</h5>
```
Comment (sethvargo) [inline assets/server/realmadmin/_stats_codes.html:120]:
```suggestion
        <h5 class="modal-title">User Report codes &amp; usage</h5>
```

File: pkg/database/token.go [hunk 2]
```diff
@@ -472,16 +480,21 @@ func (db *Database) updateStatsTokenInvalid(t time.Time, authApp *AuthorizedApp)
...
 
 	realmSQL := `
-			INSERT INTO realm_stats(date, realm_id, tokens_claimed)
-					VALUES ($1, $2, 1)
+			INSERT INTO realm_stats(date, realm_id, tokens_claimed, user_report_tokens_claimed)
+					VALUES ($1, $2, 1, $3)
 				ON CONFLICT (date, realm_id) DO UPDATE
-					SET tokens_claimed = realm_stats.tokens_claimed + 1
+					SET tokens_claimed = realm_stats.tokens_claimed + 1,
+					    user_report_tokens_claimed = realm_stats.user_report_tokens_claimed + $3
```
Comment (sethvargo) [inline pkg/database/token.go:491]: Clever

File: pkg/database/token.go [hunk 3]
```diff
@@ -472,16 +480,21 @@ func (db *Database) updateStatsTokenInvalid(t time.Time, authApp *AuthorizedApp)
 
 // updateStatsTokenClaimed updates the statistics, increasing the number of
 // tokens claimed.
-func (db *Database) updateStatsTokenClaimed(t time.Time, authApp *AuthorizedApp) {
+func (db *Database) updateStatsTokenClaimed(t time.Time, authApp *AuthorizedApp, isUserReport bool) {
```
Comment (sethvargo) [inline pkg/database/token.go:483]: I always get weary of miscellaneous booleans that control flow. Any ideas on how we could make this better? Maybe accepting a "ReportType" enum?

File: pkg/database/vercode.go [hunk 4]
```diff
@@ -327,56 +330,63 @@ func (db *Database) UpdateStats(ctx context.Context, codes ...*VerificationCode)
 
 	// If the issuer was a user, update the user stats for the day.
 	if v.IssuingUserID != 0 {
-		sql := fmt.Sprintf(`
+		sql := `
 			INSERT INTO user_stats (date, realm_id, user_id, codes_issued)
-				VALUES ($1, $2, $3, 1)
+				VALUES ($1, $2, $3, $4)
```
Comment (sethvargo) [inline pkg/database/vercode.go:335]: Ah this was a bug wasn't it.

File: pkg/database/vercode.go [hunk 5]
```diff
@@ -327,56 +330,63 @@ func (db *Database) UpdateStats(ctx context.Context, codes ...*VerificationCode)
...
 
 	// Update the per-realm stats.
 	if v.RealmID != 0 {
-		sql := fmt.Sprintf(`
-			INSERT INTO realm_stats(date, realm_id, codes_issued)
-				VALUES ($1, $2, 1)
+		// Count the number of user initiated reports
+		userReports := 0
+		for _, vc := range codes {
+			if verifyapi.ReportTypeSelfReport == vc.TestType {
```
Comment (sethvargo) [inline pkg/database/vercode.go:377]:
```suggestion
			if vc.TestType == verifyapi.ReportTypeSelfReport {
```

**Comment by sethvargo** [comment]
@mikehelmick I'm not sure I understand the "System and User Report" stats. Can you tell me more?

File: pkg/database/vercode.go (hunk 4 above)
Comment (mikehelmick) [reply pkg/database/vercode.go:335]: yes - this is a possible source of undercounting

**Comment by mikehelmick** [comment]
addressed comments + rebased

**Review by sethvargo — APPROVED** [review]

**Pull request merged**